gpii/node_modules/flowManager/src/UntrustedFlowManager.js

```diff
@@ -89,3 +85,145 @@ fluid.defaults("gpii.flowManager.untrusted.stateChangeHandler", {
+        //     accessToken: "a-token-value",
+        //     timestampExpires: "an-timestamp-the-token-expires"
+        // }
+        // ...
+    },
+    components: {
+        untrustedSettingsDataSource: {
+            type: "kettle.dataSource.URL",
+            options: {
+                // url: distributed down from the parent component gpii.flowManager.untrusted.settingsDataSource
+                termMap: {
+                    "userToken": "%userToken",
+                    "device": "%device"
+                }
+            }
+        },
+        accessRequester: {
+            type: "gpii.accessRequester",
+            options: {
+                clientCredentialDataSourceGrade: "gpii.accessRequester.clientCredentialDataSource.file"
+            }
+        }
+    },
+    invokers: {
+        get: {
+            funcName: "gpii.flowManager.untrusted.settingsDataSource.get",
+            args: ["{that}", "{arguments}.0", "{arguments}.1"]
+            // userToken, device
+        },
+        save: {
+            funcName: "gpii.flowManager.untrusted.settingsDataSource.save",
+            args: ["{that}", "{arguments}.0", "{arguments}.1", "{arguments}.2"]
+            // userToken, accessToken, timestampExpires
+        }
+    }
+});
+
+/**
+ * Retrieve user settings from the cloud using the access token requested for they keyed in user token.
+ * @param that {Component} An instance of gpii.flowManager.untrusted.settingsDataSource
+ * @param userToken {String} A user token
+ * @param device {Object} The device information provided by the device reporter
+ * @return {Promise} A promise whose resolved value is the user settings
+ */
+gpii.flowManager.untrusted.settingsDataSource.get = function (that, userToken, device) {
+    var accessTokenPromise = gpii.flowManager.untrusted.settingsDataSource.findValidAccessToken(that, userToken);
+    var promiseTogo = fluid.promise();
+
+    accessTokenPromise.then(function (accessToken) {
+        var settingsPromise = that.untrustedSettingsDataSource.get({
+            userToken: userToken,
+            device: JSON.stringify(device)
+        }, {
+            headers: {
+                "Authorization": "Bearer " + accessToken
+            }
+        });
+        fluid.promise.follow(settingsPromise, promiseTogo);
+    }, function (err) {
+        promiseTogo.reject(err);
+    });
+
+    return promiseTogo;
+};
+
+/**
+ * Find a valid access token. It first checks the saved access token for the keyed in user token,
+ * If it has expired, request and return a new one from the cloud, otherwise, return the saved access token.
+ * @param that {Component} An instance of gpii.flowManager.untrusted.settingsDataSource
+ * @param userToken {String} A user token
+ * @return {Promise} A promise whose resolved value is a valid access token
+ */
+gpii.flowManager.untrusted.settingsDataSource.findValidAccessToken = function (that, userToken) {
+    var accessToken = fluid.get(that, [userToken, "accessToken"]);
```

Comment:
It would be better to store the accessTokens inside a container in the `settingsDataSource`, rather than at the component namespace root. So that the userTokens do not clash with other names at the component namespace root. Storing the userTokens at the root of `that` means that carefully crafted userTokens could interfere with the operation of the component.